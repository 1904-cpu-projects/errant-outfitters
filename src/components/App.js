import React from "react";

class App extends React.Component {
  // gotta see that HR!
  render() {
    return (
      <div>
        <hr />
      </div>
    );
  }
}

export default App;
