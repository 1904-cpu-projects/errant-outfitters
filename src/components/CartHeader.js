import React from 'react';
import { Link } from 'react-router-dom';

// numberItems needs to just be a number
// pull in from redux store CartStore I think
export function CartHeader({ cart }) {
  const cartItems = cart ? cart.length : 0;
  return (
    <div>
      <Link to="/user/profile">Cart</Link> | {cartItems}
    </div>
  );
}
