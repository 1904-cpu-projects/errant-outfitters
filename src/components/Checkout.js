import React from 'react';
import PropTypes from 'prop-types';
import { Elements, StripeProvider } from 'react-stripe-elements';
import { connect } from 'react-redux';

import StripCard from './StripeCard';
import { getCart } from '../storeReducers/cartReducer';
import { listProductsThunk } from '../actions/productActions';

class Checkout extends React.Component {
  constructor() {
    super();
    this.state = {
      total: 0,
      transactionComplete: false,
      transacitons: [],
    };
    this.updateCart = this.updateCart.bind(this);
  }

  calcTotal(cart) {
    const total = cart.items.reduce((acc, item) => {
      return acc + item.product.cost * item.quantity;
    }, 0);
    this.setState({ total: total });
  }

  updateCart(data) {
    console.log(data);
    this.props.getCart();
    this.props.listProductsThunk();
    this.setState({ transactionComplete: true, transactions: [...data] });
  }

  componentDidMount() {
    const { cart } = this.props;
    this.calcTotal(cart);
  }

  render() {
    if (!this.state.transactionComplete) {
      return (
        <div className="checkout">
          <p>Would you like to complete the purchase?</p>
          <div>
            <StripeProvider apiKey={process.env.STRIPE_API}>
              <Elements>
                <StripCard updateCart={this.updateCart} />
              </Elements>
            </StripeProvider>
          </div>
          <div>
            <h2>Your total for this purchase is ${this.state.total}</h2>
            <p>
              You don&apos;t even need to worry about shipping information
              <br />
              We have mages on standby to teleport your items directly to you!
            </p>
          </div>
        </div>
      );
    } else {
      return (
        <div>
          <h2>Your transaction is complete!</h2>
          <div>
            Total cost: {this.state.transactions[0].totalCost}
            <br />
            Transaction #: {this.state.transactions[0].id}
            <br />
            Your Items:
            <br />
            {this.state.transactions.map((item, idx) => (
              <p key={idx}>
                {item.product.name} | Quantity: {item.quantity}
              </p>
            ))}
          </div>
        </div>
      );
    }
  }
}

Checkout.propTypes = {
  cart: PropTypes.object,
  getCart: PropTypes.func,
  listProductsThunk: PropTypes.func,
};

const mapStateToProps = state => ({
  cart: state.cart,
});

const mapDispatchToProps = dispatch => ({
  getCart: () => dispatch(getCart()),
  listProductsThunk: () => dispatch(listProductsThunk()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Checkout);
