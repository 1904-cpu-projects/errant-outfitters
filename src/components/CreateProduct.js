import React from 'react';
import { connect } from 'react-redux';
import { postProductThunk } from '../actions/productActions';
import PropTypes from 'prop-types';

class CreateProduct extends React.Component {
  constructor() {
    super();
    this.state = {
      name: '',
      description: '',
      image: '',
      category: '',
      stock: '',
      cost: '',
    };

    this.onHandle = this.onHandle.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onCategorySelection = this.onCategorySelection.bind(this);
    this.onInstockSelection = this.onInstockSelection.bind(this);
  }

  onHandle(event) {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  onCategorySelection(event) {
    this.setState({
      category: event.target.value,
    });
  }

  onInstockSelection(event) {
    this.setState({
      inStock: event.target.value,
    });
  }

  onSubmit(event) {
    event.preventDefault();
    this.props.postProductThunk(this.state);
    window.location.hash = '/#/';
  }

  render() {
    return (
      <div className="create-product-form">
        <h1>Add a new product to the set</h1>
        <form onSubmit={this.onSubmit}>
          <label htmlFor="name">Name: </label>
          <input
            onChange={this.onHandle}
            type="text"
            name="name"
            value={this.state.name}
          />

          <label htmlFor="description">Description: </label>
          <textarea
            onChange={this.onHandle}
            type="text"
            value={this.state.description}
            name="description"
          />

          <label htmlFor="image">Image: </label>
          <input
            onChange={this.onHandle}
            type="text"
            value={this.state.image}
            name="image"
          />

          <label htmlFor="inStock">inStock: </label>
          <select onChange={this.onInstockSelection} value={this.state.inStock}>
            <option value="false">false</option>
            <option value="true">true</option>
          </select>

          <label htmlFor="stock">Stock: </label>
          <input
            onChange={this.onHandle}
            name="stock"
            type="number"
            min="0"
            value={this.state.stock}
          />

          <label htmlFor="cost">Cost: </label>
          <input
            onChange={this.onHandle}
            name="cost"
            type="number"
            min="0"
            value={this.state.cost}
          />

          <label htmlFor="category">Category: </label>
          <select
            onChange={this.onCategorySelection}
            value={this.state.category}
          >
            <option key="0" value="armor">
              Armour
            </option>
            <option key="1" value="weapon">
              Weapon
            </option>
            <option key="2" value="potion">
              Potion
            </option>
          </select>
          <button type="submit">Submit</button>
        </form>
      </div>
    );
  }
}

CreateProduct.propTypes = {
  postProductThunk: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => {
  return {
    postProductThunk: thing => dispatch(postProductThunk(thing)),
  };
};

export default connect(
  null,
  mapDispatchToProps,
)(CreateProduct);
