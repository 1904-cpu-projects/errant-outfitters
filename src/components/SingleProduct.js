import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

function SingleProduct({ product }) {
  let stocked = 'No';
  if (product.inStock) {
    stocked = 'Yes';
  }
  return (
    <div key={product.id} className="product-card">
      <Link to={`/products/${product.id}`}>
        <img
          className="product-image"
          src={
            product.image ? product.image : '/img/products/default-product.jpg'
          }
          alt="Product Image"
        />
      </Link>
      <div key={product.id} className="product-container">
        <h3>
          <b> {product.name} </b>
        </h3>
        <p>
          {product.description ? product.description.slice(0, 50) + '...' : ''}
        </p>
        <p>Cost: {product.cost}</p>
        <p>In Stock: {stocked}</p> <p>Count: {product.stock}</p>
      </div>
      <Link to={`/products/${product.id}`}>
        <button>Details</button>
      </Link>
    </div>
  );
}

SingleProduct.defaultProps = {
  image: '',
  product: {},
  description: '',
  name: '',
};

SingleProduct.propTypes = {
  description: PropTypes.string,
  name: PropTypes.string,
  image: PropTypes.string,
  product: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.string,
    PropTypes.bool,
    PropTypes.number,
  ]),
};

export default SingleProduct;
