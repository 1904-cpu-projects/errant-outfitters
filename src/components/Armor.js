import React from 'react';
import { connect } from 'react-redux';
import { SingleProduct } from './SingleProduct';
import MenuBar from './MenuBar';

export function Products({ products }) {
  const armor = products.productList.filter(a => a.catagory === 'armor');
  //function to make tiles from SingleProduct component

  return (
    <div>
      <MenuBar />
      <div className="products-list">
        {armor.map(p => {
          return <SingleProduct key={p.id} product={p} />;
        })}
      </div>
    </div>
  );
}

const mapStateToProps = state => ({
  products: state.products,
});

export default connect(mapStateToProps)(Products);
