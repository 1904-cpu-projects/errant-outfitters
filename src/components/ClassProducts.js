import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { SingleProduct } from './SingleProduct';

export function Products({ products, user }) {
  const classProducts = products.productList.filter(
    a => a.class === user.class,
  );
  //function to make tiles from SingleProduct component
  if (classProducts.length !== 0) {
    return (
      <div>
        <div className="products-list">
          {classProducts.map(p => {
            return <SingleProduct key={p.id} product={p} />;
          })}
        </div>
      </div>
    );
  } else {
    return (
      <div>
        <h2 className="products-list">
          Please register/login to see your class recommended products
        </h2>
      </div>
    );
  }
}

Products.propTypes = {
  productList: PropTypes.array,
  user: PropTypes.object,
};

const mapStateToProps = state => ({
  products: state.products,
  user: state.user,
});

export default connect(mapStateToProps)(Products);
