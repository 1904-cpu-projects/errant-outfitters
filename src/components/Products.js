import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import SingleProduct from './SingleProduct';
import ClassProducts from './ClassProducts';

export function Products({ products, match }) {
  let displayFilter = [];
  if (match.params.filterBar === 'ClassProducts') return <ClassProducts />;
  match.params.filterBar === 'inStock'
    ? (displayFilter = products.productList.filter(p => p.inStock))
    : match.params.filterBar === 'outStock'
    ? (displayFilter = products.productList.filter(p => !p.inStock))
    : (displayFilter = match.params.filterBar
        ? products.productList.filter(
            p => p.catagory === match.params.filterBar,
          )
        : products.productList);

  //function to make tiles from SingleProduct component
  return (
    <div className="products-list">
      {displayFilter.map(p => {
        return <SingleProduct key={p.id} product={p} />;
      })}
    </div>
  );
}

Products.defaultProps = {
  productList: [],
  products: [],
  match: {},
};

Products.propTypes = {
  products: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object,
    PropTypes.number,
  ]),
  productList: PropTypes.arrayOf(PropTypes.object),
  match: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.number,
  ]),
};

const mapStateToProps = state => ({
  products: state.products,
});

export default connect(mapStateToProps)(Products);
