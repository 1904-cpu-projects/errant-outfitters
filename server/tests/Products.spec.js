import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow } from 'enzyme';
import { Products } from '../../src/components/Products';
import SingleProduct from '../../src/components/SingleProduct';

configure({ adapter: new Adapter() });

//first attempt at an "unconnected" component
//exported the unconnected version of the component and tested with mock data. Allows for testing the conponent in isolation from the redux store

//More tests can be written to look at other features or funcitonalities.

const mockProducts = {
  productList: [{ key1: 'value1' }, { key2: 'value2' }, { key3: 'value3' }],
};
const match = {
  params: {
    baseId: 1, //any id you want to set
  },
};
describe('<Products/>', () => {
  it('renders the Product component ', () => {
    const wrapper = shallow(<Products products={mockProducts} match={match} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('Renders a Single Product component', () => {
    const wrapper = shallow(<Products products={mockProducts} match={match} />);
    expect(wrapper.find(SingleProduct)).toHaveLength(3);
  });
});
