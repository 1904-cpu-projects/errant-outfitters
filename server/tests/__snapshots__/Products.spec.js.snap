// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products/> renders the Product component  1`] = `
<div
  className="products-list"
>
  <SingleProduct
    description=""
    image=""
    name=""
    product={
      Object {
        "key1": "value1",
      }
    }
  />
  <SingleProduct
    description=""
    image=""
    name=""
    product={
      Object {
        "key2": "value2",
      }
    }
  />
  <SingleProduct
    description=""
    image=""
    name=""
    product={
      Object {
        "key3": "value3",
      }
    }
  />
</div>
`;
